### ISSUE TITLE
Add option to burn subtitles

### ISSUE THREAD

**Issue by LetMeByte** [crunchy-labs/crunchy-cli#140]

**Comment by ByteDream** [comment]
What subtitles should be burned / on which command? Could you specify your question/request a bit more?

**Comment by LetMeByte** [comment]
I'm sorry if my request was ambiguous.

What I'm thinking about is an option like `--burn-subs` or `--hardsub`, which would burn subtitles into the video stream via [ffmpeg's `-vf subtitles` filter](https://trac.ffmpeg.org/wiki/HowToBurnSubtitlesIntoVideo). It would make sense for it to throw an error if multiple subtitle languages are selected for download.

**Comment by ByteDream** [comment]
Isn't this covered by the `download` command?

**Comment by LetMeByte** [comment]
Not as far as I know, the `download` option does add the subtitles as softsubs into the container, though. Some players don't support softsubs, so burning hardsubs would be useful there.

**Comment by ByteDream** [comment]
Okay I see. Your suggested ffmpeg options are already implemented [here](https://github.com/crunchy-labs/crunchy-cli/blob/96b259ce9a0fbc7678195dbca6742ade5986f7e0/crunchy-cli-core/src/cli/download.rs#L340-L343) but isn't available in a pre-release yet. It will be available in the next one.

**Comment by LetMeByte** [comment]
Oh well, I didn't read the source this time. I'm happy it's already done, thanks for the great work!

**Comment by ByteDream** [comment]
> Edit: Does that solution burn subtitles by default? I could not determine that by reading the code, but I think burning subtitles should be optional.

Yes it does. The desired behavior for subtitles in `download` is to be burnt-in. For everything softsub related, `archive` should be used.

**Comment by LetMeByte** [comment]
Right, didn't think about it like that.

But one more thing; the problem [here](https://github.com/crunchy-labs/crunchy-cli/blob/96b259ce9a0fbc7678195dbca6742ade5986f7e0/crunchy-cli-core/src/cli/download.rs#L325-L326) is happening because you can't burn subs without re-encoding the video stream, so you need to specify a codec (e.g. `-c:v h264_nvenc`) instead of just copying the video stream with `-c:v copy`. Without specifiyng a codec ffmpeg will determine the supposed 'best' option, which does not take hardware acceleration into account. Sorry in advance if that is also already solved, just caught my eye.

**Comment by ByteDream** [comment]
> Therefore substituting -c:v copy with the codec provided by the --ffmpeg-preset profile, e.g. -c:v h264_nvenc for h264-nvidia-lossless, would entirely solve this problem.

This is already done. The `output_presets` variable (where `-c:v copy` gets removed from) is generated from the `--ffmpeg-preset` flag. `-c:v` has `copy` only as value if no hardware acceleration is provided. If `--ffmpeg-preset` is set to `h264-nvidia-lossless` (for example), `output_presets` contains `-c:v h264_nvenc` instead of `-c:v copy` and therefore nothing gets touched/removed.

> Using the -vf ass=subtitles.ass filter instead of -vf subtitles=subtitle.ass would also be more 'correct', as CR is using ASS formatted subtitles anyway.

Thanks for mentioning that, didn't knew this specific flag exists.

**Comment by LetMeByte** [comment]
> If --ffmpeg-preset is set to h264-nvidia-lossless (for example), output_presets contains -c:v h264_nvenc instead of -c:v copy and therefore nothing gets touched/removed.

Now that you said that, I do understand that part of the code. I'm not much of a programmer, so please excuse me misreading the code (or not understanding it at all).

> Thanks for mentioning that, didn't know this specific flag exists.

No problem, it doesn't really matter though, ffmpeg would've done the right thing anyway.

What about the audio though? As far as I know ffmpeg will never implicitly convert aac (or any audio?) to a different format, therefore the `-c:a copy` flag can always be omitted.

**Comment by ByteDream** [comment]
> Now that you said that, I do understand that part of the code. I'm not much of a programmer, so please excuse me misreading the code (or not understanding it at all).

No problem, I'm always happy to help and to make it a bit more understandable :)

> What about the audio though? As far as I know ffmpeg will never implicitly convert aac (or any audio?) to a different format, therefore the -c:a copy flag can always be omitted.

Is there any source for this? I've skipped through the documentation (and google a bit) but didn't found anything.

**Comment by LetMeByte** [comment]
Ok, I found something. [The docs](https://ffmpeg.org/ffmpeg.html#Stream-handling) state that:
> If no `-codec` option is specified for a stream type, ffmpeg will select the default encoder registered by the output file muxer.

I know for a fact that the default audio encoder for the mp4 output file muxer is ffmpeg's native `aac` encoder, so it won't encode CR's audio if we save it to an `.mp4` file.

I have to check for `.mkv` though, and I want a list of default encoders.
